HPXMLtoOpenStudio/tests/test_validation.rb

```diff
@@ -0,0 +1,618 @@
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/WaterHeating/WaterHeatingSystem[WaterHeaterType="instantaneous water heater"]'
+      return 'base-dhw-tankless-electric.xml'
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/WaterHeating/WaterHeatingSystem[WaterHeaterType="heat pump water heater"]'
+      return 'base-dhw-tank-heat-pump.xml'
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/WaterHeating/WaterHeatingSystem[WaterHeaterType="space-heating boiler with storage tank"]'
+      return 'base-dhw-indirect.xml'
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/WaterHeating/WaterHeatingSystem[WaterHeaterType="space-heating boiler with tankless coil"]/RelatedHVACSystem'
+      return 'base-dhw-combi-tankless.xml'
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/WaterHeating/WaterHeatingSystem[UsesDesuperheater="true"]'
+      return 'base-dhw-desuperheater.xml'
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/WaterHeating/HotWaterDistribution/SystemType/Recirculation'
+      return 'base-dhw-recirc-timer.xml'
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/WaterHeating/HotWaterDistribution/DrainWaterHeatRecovery'
+      return 'base-dhw-dwhr.xml'
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/SolarThermal/SolarThermalSystem[SolarFraction]'
+      return 'base-dhw-solar-fraction.xml'
+    elsif ['/HPXML/Building/BuildingDetails/Systems/SolarThermal/SolarThermalSystem',
+           '/HPXML/Building/BuildingDetails/Systems/SolarThermal/SolarThermalSystem[CollectorArea]'].any? { |i| key.include? i }
+      return 'base-dhw-solar-direct-flat-plate.xml'
+    elsif key.include? '/HPXML/Building/BuildingDetails/Systems/Photovoltaics/PVSystem'
+      return 'base-pv.xml'
+    elsif ['/HPXML/Building/BuildingDetails/Appliances/Freezer', '/HPXML/Building/BuildingDetails/Pools/Pool',
+           '/HPXML/Building/BuildingDetails/HotTubs/HotTub', '/HPXML/Building/BuildingDetails/MiscLoads/FuelLoad'].any? { |i| key.include? i }
+      return 'base-misc-large-uncommon-loads.xml'
+    else
+      return 'base.xml'
+    end
+  end
+
+  def test_valid_sample_files
+    sample_files_dir = File.absolute_path(File.join(@root_path, 'workflow', 'sample_files'))
+    hpxmls = []
+    Dir["#{sample_files_dir}/*.xml"].sort.each do |xml|
+      hpxmls << File.absolute_path(xml)
+    end
+    puts "Testing #{hpxmls.size} HPXML files..."
+    hpxmls.each do |hpxml_path|
+      print '.'
+      # Schematron validation
+      _test_schematron_validation(hpxml_path)
+
+      # .rb validator validation
+      hpxml = HPXML.new(hpxml_path: File.join(@root_path, 'workflow', 'sample_files', File.basename(hpxml_path)))
+      hpxml_doc = hpxml.to_oga()
+      _test_ruby_validation(hpxml_doc)
+      # Schematron validation
+      _test_schematron_validation(hpxml_path)
+    end
+    puts
+  end
+
+  def test_invalid_files
+    expected_error_msgs = {
```

Comment:
Call this expected_error_msgs_required, to contrast with expected_error_msgs_optional below.